Source/santa-driver/CircularQueue/Queue.c

```diff
@@ -0,0 +1,652 @@
+    // should be inserted.
+    HELM_ADVANCE_WRAP,
+    // Failure: there is insufficient room for the write window to expand this
+    // much. Can be retried.
+    HELM_ADVANCE_NO_ROOM,
+    // Permanent failure: The requested buffer is too large to EVER fit on this
+    // queue. Do not retry.
+    HELM_ADVANCE_TOO_LARGE,
+} helm_advance_ret_t;
+
+// Returns the next logical slot, wrapping around as necessary.
+// Caller MUST ensure that the memory containing the next slot is initialized!
+static inline helm_queue_slot_t *next_slot(helm_queue_t *queue,
+                                           helm_queue_slot_t *slot,
+                                           size_t limit) {
+    size_t off_slot = reloffset(queue, slot);
+    size_t off_next = off_slot + helm_queue_slot_size(slot);
+    
+    // Don't pass the limit, which may be below us.
+    if ((off_slot <= limit && off_next >= limit)) {
+        return NULL;
+    }
+    
+    // At this point it's correct to have a pointer to the memory where the next
+    // slot might be.
+    helm_queue_slot_t *next = get_slot(queue, off_next);
+    
+    HELM_BUG_ON(off_next + helm_queue_slot_size(next) > queue->buffer_size,
+                "Next slot would exceed the allocated buffer.");
+    
+    if (next->flags & HELM_QUEUE_SLOT_WRAPAROUND) {
+        return first_slot(queue);
+    }
+    
+    return next;
+}
+
+// Looks for a way to fit a contiguous buffer of 'reqsize' bytes onto the queue.
+// On success, sets 'cursor' to the new value of write_window.top. Caller MUST
+// ensure the queue is locked. See helm_advance_ret_t for return values.
+static inline helm_advance_ret_t advance_write_cursor(helm_queue_t *queue,
+                                                      size_t reqsize,
+                                                      size_t *cursor) {
+    *cursor = 0;
+    
+    if (reqsize > usable_buffer_size(queue)) {
+        return HELM_ADVANCE_TOO_LARGE;
+    }
+    
+    if ((distance(queue, queue->write_window.top, queue->read_window.bottom) >
+         reqsize) ||
+        is_empty(queue)) {
+        // There is enough room to fit reqsize under the read window without the top
+        // of the write window exceeding or touching the bottom of the read window.
+        if (queue->write_window.top + reqsize <= usable_buffer_size(queue)) {
+            // There is enough room without wrapping.
+            *cursor = queue->write_window.top + reqsize;
+            return HELM_ADVANCE_ADVANCE;
+        }
+        
+        if (reqsize < queue->read_window.bottom) {
+            // There is enough room under the read window at the start of the buffer,
+            // so we wrap around.
+            *cursor = 0;
+            return HELM_ADVANCE_WRAP;
+        }
+        
+        // If we're here then there is enough room on the buffer but it's not in one
+        // contiguous chunk. Too bad - we need to wait.
+    }
+    
+    return HELM_ADVANCE_NO_ROOM;
+}
+
+#ifdef NDEBUG
```

Comment:
We have been using DEBUG throughout since it is defined by default with the xcode tools.